luigi/target.py

```diff
@@ -257,14 +251,62 @@ def remove(self):
         """
         Remove the resource at the path specified by this FileSystemTarget.
 
-        This method is implemented by using :py:meth:`fs`.
+        This method is implemented by using :py:attr:`fs`.
         """
         self.fs.remove(self.path)
 
+    def temporary_path(self):
+        """
+        A context manager that enables a reasonably short, general and
+        magic-less way to solve the :ref:`AtomicWrites`. When the context
+        manager enters it will not do anything, but when it exits it will move
+        the file if there was no exception thrown.
+
+        The final move operation will be carried out by calling
+        :py:meth:`FileSystem.rename_dont_move` on :py:attr:`fs`.
+
+        The typical use case looks like this:
+
+        .. code:: python
+
+            class MyTask(luigi.Task):
+                def output(self):
+                    return MyFileSystemTarget(...)
+
+                def run(self):
+                    with self.output().temporary_path() as self.temp_output_path:
+                        run_some_external_command(output_dir=self.temp_output_path)
+        """
+        class _Manager(object):
+            target = self
+
+            def __init__(self):
+                num = random.randrange(0, 1e10)
+                self._temp_path = '{}-luigi-tmp-{:010}{}'.format(
+                    self.target.path.rstrip('/'),
+                    num,
+                    '/' if self.target._has_trailing_slash() else '')
+
+            def __enter__(self):
+                return self._temp_path
+
+            def __exit__(self, exc_type, exc_value, traceback):
+                if exc_type is None:
+                    # There were no exceptions
+                    self.target.fs.rename_dont_move(self._temp_path, self.target.path)
+                return False  # False means we don't suppress the exception
```

Comment:
I've thought about this actually. And I think it's best to not clean them up. I usually find the temporary files to be invaluable traces left when debugging. I'm however up for in an upcoming patch add a kwarg like `delete_on_failure=False`.